feat: Migrate sample and sample tests - Batch 2 of 5

vinnysenthil opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-aiplatform/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #9 (Batch 2 of 5) 🦕

leahecole commented on samples/tests/create_endpoint_test.py:
```diff
@@ -0,0 +1,53 @@
+
+DISPLAY_NAME = f"temp_create_endpoint_test_{uuid4()}"
+PROJECT = "ucaip-sample-tests"
+ENDPOINT_ID = None
```
```suggestion
pytest.ENDPOINT_ID = None
```

vinnysenthil replied:
Replaced mutable global vars with `pytest` fixtures

leahecole commented on samples/tests/create_endpoint_test.py:
```diff
@@ -0,0 +1,53 @@
+PROJECT = "ucaip-sample-tests"
+ENDPOINT_ID = None
+
+@pytest.fixture(scope="function", autouse=True)
+def teardown(capsys):
+    yield
+
+    assert ENDPOINT_ID is not None
+
+    # Delete the endpoint that was just created
+    delete_endpoint_sample.delete_endpoint_sample(
+        project=PROJECT,
+        endpoint_id=ENDPOINT_ID
+    )
+
+    out, _ = capsys.readouterr()
+    assert "delete_endpoint_response:" in out
+
```
```suggestion
@pytest.fixture(scope="function", autouse=True)
def teardown():
    yield

    assert pytest.ENDPOINT_ID is not None

    # Delete the endpoint that was just created
    delete_endpoint_sample.delete_endpoint_sample(
        project=PROJECT,
        endpoint_id=pytest.ENDPOINT_ID
    )

```

leahecole replied:
I removed the assertion in the fixture, because it's not a test, but if you are concerned about the operation failing, you can add a try/except block

vinnysenthil replied:
Removed assertion from fixture

leahecole commented on samples/tests/create_endpoint_test.py:
```diff
@@ -0,0 +1,53 @@
+    assert "delete_endpoint_response:" in out
+
+
+def test_ucaip_generated_create_endpoint_sample(capsys):
+    global ENDPOINT_ID
+    assert ENDPOINT_ID is None
+
+    create_endpoint_sample.create_endpoint_sample(
+        display_name=DISPLAY_NAME,
+        project=PROJECT
+    )
+
+    out, _ = capsys.readouterr()
+    assert "create_endpoint_response" in out
+
+    ENDPOINT_ID = out.split("name:")[1].split("\n")[0].split("/")[-1]
```
```suggestion
def test_ucaip_generated_create_endpoint_sample(capsys):
    assert pytest.ENDPOINT_ID is None

    create_endpoint_sample.create_endpoint_sample(
        display_name=DISPLAY_NAME,
        project=PROJECT
    )

    out, _ = capsys.readouterr()
    assert "create_endpoint_response" in out

   pytest.ENDPOINT_ID = out.split("name:")[1].split("\n")[0].split("/")[-1]
```

vinnysenthil replied:
Replaced mutable global vars with `pytest` fixtures

leahecole commented on samples/tests/create_training_pipeline_image_classification_test.py:
```diff
@@ -0,0 +1,67 @@
+DATASET_ID = "1084241610289446912"  # Permanent 50 Flowers Dataset
+DISPLAY_NAME = f"temp_create_training_pipeline_image_classification_test_{uuid4()}"
+
+TRAINING_PIPELINE_NAME = None
```
```suggestion
pytest.TRAINING_PIPELINE_NAME = None
```

vinnysenthil replied:
Replaced mutable global vars with `pytest` fixtures

leahecole commented on samples/tests/create_training_pipeline_image_classification_test.py:
```diff
@@ -0,0 +1,67 @@
+TRAINING_PIPELINE_NAME = None
+
+
+@pytest.fixture(scope="function", autouse=True)
+def teardown(capsys):
+    yield
+
+    assert TRAINING_PIPELINE_NAME is not None
+
+    training_pipeline_id = TRAINING_PIPELINE_NAME.split("/")[-1]
+
+    # Stop the training pipeline
+    cancel_training_pipeline_sample.cancel_training_pipeline_sample(
+        project=PROJECT_ID, training_pipeline_id=training_pipeline_id
+    )
+
+    # Delete the training pipeline
+    delete_training_pipeline_sample.delete_training_pipeline_sample(
+        project=PROJECT_ID, training_pipeline_id=training_pipeline_id
+    )
+
+    out, _ = capsys.readouterr()
+    assert "delete_training_pipeline_response" in out
```
```suggestion
@pytest.fixture(scope="function", autouse=True)
def teardown():
    yield

    assert pytest.TRAINING_PIPELINE_NAME is not None

    training_pipeline_id = pytest.TRAINING_PIPELINE_NAME.split("/")[-1]

    # Stop the training pipeline
    cancel_training_pipeline_sample.cancel_training_pipeline_sample(
        project=PROJECT_ID, training_pipeline_id=training_pipeline_id
    )

    # Delete the training pipeline
    delete_training_pipeline_sample.delete_training_pipeline_sample(
        project=PROJECT_ID, training_pipeline_id=training_pipeline_id
    )
```

vinnysenthil replied:
Removed assertions from fixtures

leahecole commented on samples/tests/create_training_pipeline_image_classification_test.py:
```diff
@@ -0,0 +1,67 @@
+    assert "delete_training_pipeline_response" in out
+
+
+def test_ucaip_generated_create_training_pipeline_video_classification_sample(capsys):
+    global TRAINING_PIPELINE_NAME
+    assert TRAINING_PIPELINE_NAME is None
+
+    create_training_pipeline_image_classification_sample.create_training_pipeline_image_classification_sample(
+        project=PROJECT_ID,
+        display_name=DISPLAY_NAME,
+        dataset_id=DATASET_ID,
+        model_display_name=f"Temp Model for {DISPLAY_NAME}",
+    )
+
+    out, _ = capsys.readouterr()
+
+    # Save resource name of the newly created training pipeline
+    TRAINING_PIPELINE_NAME = out.split("name:")[1].split("\n")[0]
```
```suggestion
def test_ucaip_generated_create_training_pipeline_video_classification_sample(capsys):
    assert pytest.TRAINING_PIPELINE_NAME is None

    create_training_pipeline_image_classification_sample.create_training_pipeline_image_classification_sample(
        project=PROJECT_ID,
        display_name=DISPLAY_NAME,
        dataset_id=DATASET_ID,
        model_display_name=f"Temp Model for {DISPLAY_NAME}",
    )

    out, _ = capsys.readouterr()

    # Save resource name of the newly created training pipeline
    pytest.TRAINING_PIPELINE_NAME = out.split("name:")[1].split("\n")[0]
```

leahecole replied:
There should be an assertion in this test based on the output being captured - even if the operation hasn't finished, you can assert that there is an operation that's related to what you just created - I know you had something in the fixture before, but fixtures are just for setting up/tearing down the resources, not for the actual test logic

vinnysenthil replied:
Added assertions to all tests

leahecole commented on samples/create_training_pipeline_image_classification_sample.py:
```diff
@@ -0,0 +1,164 @@
+        parent=parent, training_pipeline=training_pipeline
+    )
+    print("response")
+    print(" name:", response.name)
```
it seems like theres a lot of print statements throughout this sample - is printing all of these things beneficial to the user? If it's a lot of info, would it possibly be more valuable to print the output to some kind of file or to gcs? Y'all know what's best for uCAIP, but I'm just curious about seeing `print` so much here :)

dinagraves replied:
+1, if we want to print out the data, maybe use some kind of method like outlined here: https://stackoverflow.com/a/59128615/12375776

leahecole commented on CHANGELOG.md:
```diff
@@ -1 +1,9 @@
 # Changelog
```
Is this autogenerated? Should ti be included?

busunkim96 replied:
I think this is from merging into the release branch`release-v0.2.0`. You should be able to change the base branch back to `master`.

leahecole reviewed (changes requested):
I didn't go super in depth to each of them, but looked at the first three and found some common changes that I suspect are going to be common throughout

- Avoid using globals, and instead emulate the pattern found in this [dialogflow test](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/dialogflow/cloud-client/create_knowledge_base_test.py) for passing the variable to the fixture
- Avoid testing the output of a fixture - keep your test logic to your tests. Instead, in the fixture, if you're worried about an operation failing or not being able to occur, utilize try/excepts (i.e., try to delete an endpoint, except notfound, print already deleted)
- Make sure you do have an assertion based on the output in the test itself
- Do there need to be as many print statements as there are in the samples, or is there a prettier way to get the user the info they need? (Either less info, or writing it to a file?) - this is the question I feel least strongly about and am inclined to trust you uCAIP folks

busunkim96 commented on samples/create_endpoint_sample.py:
```diff
@@ -0,0 +1,39 @@
+    endpoint = {"display_name": display_name}
+    response = client.create_endpoint(parent=parent, endpoint=endpoint)
+    print("Long running operation:", response.operation.name)
+    create_endpoint_response = response.result(timeout=300)
```
If the default timeout is too low I would also edit the gRPC service config.

leahecole replied:
I still stand by this comment - does the gRPC service config need to be changed?

dizcology replied:
This is a little tricky.  In a new project (where an endpoint has not been created), the first time this method is called it could take up to 300 seconds, but subsequent calls in the same project finishes within seconds.  Since GAPIC library internally polls the long running operation, the tests typically do not run that long.

Agreed that perhaps it's better to change the timeout at the gRPC service config.

leahecole reviewed (changes requested):
Can you ensure that the sample names and the test names match? I'm guessing that create_training_pipeline_tabular_regression_sample and create_training_pipeline_tables_regression_test are related, as well as create_training_pipeline_tabular_classification_sample and create_training_pipeline_tables_classification_test

Unless I'm wrong and those are different in which case it seems like the Prs might be a little mixed up


Additionally, your tests are missing assertions entirely. You don't want them to be in the fixture, but you do want to make sure that you are making some assertions in the tests themselves

vinnysenthil commented:
Hi @leahecole - I've addressed requested changes:

- Moved to using fixtures instead of mutable globals
- Ensured assertions are made in every test
- Got the filenames for tests + samples synced
- Other changes to tests to get the passing, tests should successfully run in `sample-final-stage` once all five batch PRs are merged.

#11 has been merged which contains `helpers.py`. Merging the rest of the PRs in order

leahecole commented on samples/tests/create_training_pipeline_image_classification_test.py:
```diff
@@ -0,0 +1,82 @@
+    )
+
+    out, _ = capsys.readouterr()
+    assert "response:" in out
```
if there's an error or the response is blank, will this pass and give a false positive? If there's something deeper you can assert, or an additional assertion you can add, that will make this stronger

aribray replied:
It'll be more efficient to go over the testing assertions once all of the branches are merged into one, so I'm going to resolve this now and then address these once we can test

leahecole commented on samples/create_training_pipeline_image_classification_sample.py:
```diff
@@ -0,0 +1,54 @@
+    training_task_inputs_dict = {
+        "multiLabel": True,
+        "modelType": "CLOUD",
+        "budgetMilliNodeHours": 8000,
```
to my untrained eye this is a big scary number - will it cost the customer lots of money? Is there a reason for choosing this number in particular?

vinnysenthil replied:
TODO: Add link to docs for node hours

leahecole commented on samples/deploy_model_sample.py:
```diff
@@ -0,0 +1,54 @@
+    model_name: str,
+    deployed_model_display_name: str,
+    location: str = "us-central1",
+    timeout: int = 7200,
```
does this timeout need to be this high?

aribray replied:
Same idea as the other comment - I'm going to create issues for these, for further discussion

leahecole reviewed (changes requested):
nit for all - direcotry reorg doesn't have to happen yet, but test names should match sample names, just with `_test` on the end :) 

With some of the assertions, you are only looking for response - I want to double check that they won't give you a false positive if the response is blank/an error - double check that response is only going to happen if it for realz passes, and if not, adding another assertion to make it stronger

In some samples, you have a specific number of node hours declared - is this something that will be clear to the user? are these numbers arbitrary? if it's not going to be specified in accompanying docs, might be worth adding some comments, especially if these are decisions that will affect billing

aribray commented on samples/create_endpoint_sample.py:
```diff
@@ -0,0 +1,34 @@
+
+
+def create_endpoint_sample(
+    project: str, display_name: str, location: str = "us-central1", timeout: int = 300
```
Talked to Vinny, and we're going to create separate issues for these.

leahecole commented:
approving and merging to next stage despite test failures - no test failures when we go to master though :)
